src/include/connection.h

```diff
@@ -42,6 +42,20 @@ struct __wt_process {
 };
 extern WT_PROCESS __wt_process;
 
+/*
+ * WT_BACKGROUND_COMPACT --
+ *	Structure dedicated to the background compaction server
+ */
+struct __wt_background_compact {
+    WT_SPINLOCK cfg_lock;     /* Configuration lock */
+    WT_SESSION_IMPL *session; /* Thread session */
+    WT_CONDVAR *cond;         /* Wait mutex */
+    wt_thread_t tid;          /* Thread */
+    volatile bool running;    /* Compaction supposed to run */
```

Comment:
So, PM-3221 actually advocates for the usage of volatile (but in general `READ_ONCE` semantics). We like it for:
- Introduced read prevention
- Read tearing prevention (less of an issue)
- Compiler optimization prevention

In the case I see in this PR, we certainly need `volatile` as it prevents the compiler from optimizing out the if statements where we read it. If we were to lock around the places we read the `running` field we would achieve the same outcome. Now `volatile` doesn't force the CPU to do anything so if we are worried that the CPU won't do what we want, e.g. read a value from its cache. Then we should use locks, that will avoid needing to think about this.

Reply:
Thank you both for pointing this out!
I have tried an approach with locks, see 0925f2015adbc2458f57856374de5e507015e8b4.
I am not a big fan as I feel that I might be using those locks too much, let me know what you both think.

Reply:
Since @luke-pearson is happy to sanction this style of `volatile` usage, I'm OK sticking with the original implementation. It will be consistent with what we do for other background threads and if we want to reconsider in the future, we can make a whole sale change across all of them.

@etienneptl, I'm not sure what specific concern you have when you say it may be "using those locks too much." I don't think it uses the locks more than required for synchronization and correctness. And I don't think this additional locking would cause performance problems because the locks will seldom be contended, we don't hold them for long, and the places we use them aren't performance critical.

Reply:
> @etienneptl, I'm not sure what specific concern you have when you say it may be "using those locks too much." I don't think it uses the locks more than required for synchronization and correctness. And I don't think this additional locking would cause performance problems because the locks will seldom be contended, we don't hold them for long, and the places we use them aren't performance critical.

After re-reading the changes, I agree with you. I was not convinced at first but it seems fine in the end!